firewood/src/proof.rs

```diff
@@ -102,11 +102,95 @@ const BRANCH_NODE_SIZE: usize = 17;
 /// to a single proof step. If reaches an end step during proof verification,
 /// the hash value will be none, and the encoded value will be the value of the
 /// node.
-pub struct SubProof {
+struct SubProof {
     encoded: Vec<u8>,
     hash: Option<[u8; 32]>,
 }
 
+fn generate_subproof(data: Vec<u8>) -> Result<SubProof, ProofError> {
+    match data.len() {
+        0..=31 => {
+            let sub_hash = sha3::Keccak256::digest(&data).into();
+            Ok(SubProof {
+                encoded: data,
+                hash: Some(sub_hash),
+            })
+        }
+
+        32 => {
+            let sub_hash: &[u8] = &data;
+            let sub_hash = sub_hash.try_into().unwrap();
+
+            Ok(SubProof {
+                encoded: data,
+                hash: Some(sub_hash),
+            })
+        }
+
+        len => Err(ProofError::DecodeError(Box::new(
+            bincode::ErrorKind::Custom(format!("invalid proof length: {len}")),
+        ))),
+    }
+}
+
+fn locate_subproof<'a>(
+    mut key_nibbles: NibblesIterator<'a, 0>,
+    encoded_node: &[u8],
+) -> Result<(Option<SubProof>, NibblesIterator<'a, 0>), ProofError> {
+    let items: Vec<Encoded<Vec<u8>>> = bincode::DefaultOptions::new()
+        .deserialize(encoded_node)
+        .map_err(ProofError::DecodeError)?;
```

Comment:
This loses the reason why the deserialize fails, although I'm not sure that's easy to fix.

Reply:
Did you look at the deleted code? I haven't changed anything. Also, I don't think we lose any information; `ProofError::DecodeError` contains the `bincode::Error`.